### PR TITLE
Sf movies 3

### PR DESCRIPTION
**Pull request by vinay-pad**
What: Get movies using filters by doing a  GET to /movies endpoint

Why: These changes are part of the third part of how to build your own api, involving getting results from the created movie database.

Details:
This checkin involves following changes in order to be get movie results.

- Use the hapi-query-filter dependency to be able to do db queries to the movies database.
- Add queries for getting movies by title, release_year, release year range, exact title and fuzzy title.
- Made use of pg_trgm (trigram) module for postgres to be able to perform fuzzy queries using the '%' operator.

### CODE REVIEW

File: test/plugins/features/movies/index.test.js [hunk 1]
```diff
@@ -20,4 +21,86 @@ describe('movies integration', () => {
 
   });
 
+  describe('get', () => {
```
Comment (thetonywu) [inline test/plugins/features/movies/index.test.js]: nit: 'list'

File: test/plugins/features/movies/index.test.js [hunk 2]
```diff
@@ -20,4 +21,86 @@ describe('movies integration', () => {
...
+
+    it('Gets all movies', () => {
```
Comment (thetonywu) [inline test/plugins/features/movies/index.test.js]: these are all good tests, but it looks like you're putting all the tests at the integration level and not the controller level. Most of these should go in controller.test.js, and then you can just one test per method (create/list) in this file to simulate the end to end API calls

File: .eslintrc [hunk 3]
```diff
@@ -2,5 +2,8 @@
   extends: "lob",
   globals: {
     expect: false
+  },
+  rules: {
```
Comment (thetonywu) [inline .eslintrc]: why did you add this?

**Review by thetonywu — COMMENTED** [review]
a few comments

File: .eslintrc (hunk 3 above)
Comment (vinay-pad) [reply .eslintrc]: I included "function" keyword before "(qb)" which was throwing me a lint error which is why i excluded this rule. But i realized the mistake and have fixed this now.

File: test/plugins/features/movies/controller.test.js [hunk 4]
```diff
@@ -23,5 +24,55 @@ describe('movie controller', () => {
 
   });
 
+  describe('list', () => {
+
+    beforeEach(() => {
+      return Knex.raw('TRUNCATE movies CASCADE;');
+    });
+
+    beforeEach(() => {
```
Comment (thetonywu) [inline test/plugins/features/movies/controller.test.js]: you can combine these two beforeEach blocks and use .then to chain them

`
beforeEach(() => {
     return Knex.raw('TRUNCATE movies CASCADE;')
     .then(() => {
        return Knex('movies').insert(movie_list);
     });
 });
`

**Review by thetonywu — APPROVED** [review]
one comment, lgtm

File: lib/plugins/features/movies/controller.js [hunk 5]
```diff
@@ -8,3 +8,25 @@ exports.create = (payload) => {
     return new Movie({ id: movie.id }).fetch();
   });
 };
+
+exports.findAll = (filter) => {
+  return new Movie()
+  .query((qb) => {
+    if (filter.year !== undefined) {
```
Comment (itsachen) [inline lib/plugins/features/movies/controller.js]: Nit: `undefined` is [falsy](https://developer.mozilla.org/en-US/docs/Glossary/Falsy#Examples), so you could be more succinct by checking against the value:
```js
if (filter.year) {
  ...
}
```

File: lib/plugins/features/movies/index.js [hunk 6]
```diff
@@ -16,6 +17,23 @@ exports.register = (server, options, next) => {
       }
     }
   }]);
+  server.route([{
```
Comment (itsachen) [inline lib/plugins/features/movies/index.js]: Since [`server.route`](https://hapijs.com/api#serverrouteoptions) accepts an array of objects, you should append the configuration for GET movies to the original `server.route` invocation.

File: test/plugins/features/movies/controller.test.js [hunk 7]
```diff
@@ -23,5 +24,55 @@ describe('movie controller', () => {
...
+    ];
+
+    beforeEach(() => {
+      return Knex.raw('TRUNCATE movies CASCADE;')
+        .then(() => {
+          return Knex('movies').insert(movie_list);
+        });
+    });
+
+    it('List all movies. No filters specified', () => {
```
Comment (itsachen) [inline test/plugins/features/movies/controller.test.js]: Nit: Usually we name tests using TDD-inspired syntax so that when the full test name is constructed, based on the nested `it`'s and `describe`'s, the name reads like one sentence.

In this case if the test were renamed `lists all movies with no filters specified`, the full test name would be `list lists all moves with no filters specified`. See [api](https://github.com/lob/lob-api/blob/master/test/plugins/features/areas/index.test.js#L73) for further examples.

File: test/plugins/features/movies/controller.test.js [hunk 8]
```diff
@@ -23,5 +24,55 @@ describe('movie controller', () => {
...
+        release_year: 2012
+      },
+      {
+        title: 'Bedazzled',
+        release_year: 2007
+      }
+    ];
+
+    beforeEach(() => {
+      return Knex.raw('TRUNCATE movies CASCADE;')
```
Comment (itsachen) [inline test/plugins/features/movies/controller.test.js]: Nit: Not sure if the semicolon is necessary

File: test/plugins/features/movies/index.test.js [hunk 9]
```diff
@@ -1,16 +1,22 @@
...
 
 const Movies = require('../../../../lib/server');
+const Knex   = require('../../../../lib/libraries/knex');
 
 describe('movies integration', () => {
 
   describe('create', () => {
 
+    beforeEach(() => {
+      return Knex.raw('TRUNCATE movies CASCADE;');
```
Comment (itsachen) [inline test/plugins/features/movies/index.test.js]: Nit: Not sure if the semicolon is required

File: test/plugins/features/movies/index.test.js [hunk 10]
```diff
@@ -1,16 +1,22 @@
...
+    });
+
     it('creates a movie', () => {
       return Movies.inject({
         url: '/movies',
         method: 'POST',
-        payload: { title: 'Volver' }
+        payload: { title: 'Volver',
+		   release_year: '2011' }
```
Comment (itsachen) [inline test/plugins/features/movies/index.test.js]: Super nit: align either the colons or start of the keys in the new line

File: test/plugins/features/movies/index.test.js [hunk 11]
```diff
@@ -20,4 +26,80 @@ describe('movies integration', () => {
...
+        expect(response.result[0]).to.have.all.keys([
+          'id',
+          'title',
+          'release_year',
+          'object'
+        ]);
+      });
+    });
+
+    it('Lists movies for a given range of release years', () => {
```
Comment (itsachen) [inline test/plugins/features/movies/index.test.js]: Nit: Its not a hard and fast rule but normally I'd expect tests that check most of the functionality (different filter methods) of the controller to be in the controller unit tests and for integration tests to handle more of the end-to-end testing of the endpoint (validation, authentication, error codes).

File: .travis.yml [hunk 12]
```diff
@@ -8,6 +8,7 @@ before_script:
   - npm run db:setup:user
   - npm run db:reset
 script:
+  - psql --dbname sfmovies_test --command "CREATE EXTENSION pg_trgm;"
```
Comment (itsachen) [inline .travis.yml:11]: Question: Is this required for fuzzy search?

**Review by itsachen — APPROVED** [review]
LGTM, just some nits and comments

File: .travis.yml (hunk 12 above)
Comment (itsachen) [reply .travis.yml:11]: Nevermind, just read the PR description 😄

File: test/plugins/features/movies/index.test.js (hunk 11 above)
Comment (itsachen) [reply test/plugins/features/movies/index.test.js]: Looks like @tonywuhl had similar thoughts

File: test/plugins/features/movies/controller.test.js (hunk 7 above)
Comment (vinay-pad) [reply test/plugins/features/movies/controller.test.js]: Thanks for the tip! 👍

File: test/plugins/features/movies/index.test.js (hunk 11 above)
Comment (vinay-pad) [reply test/plugins/features/movies/index.test.js]: Yup i fixed this to have all the cases in the controller tests and only end to end tests in integration test.
